File: python/proto-macro-tool/proto_macro_tool.ml
let read_file (filename : string) : string =
  In_channel.with_open_bin filename (fun input -> In_channel.input_all input)

let write_file (filename : string) (contents : string) : unit =
  Out_channel.with_open_bin filename (fun output ->
      Out_channel.output_string output contents)

let copy_file (src : string) (dst : string) : unit =
  write_file dst (read_file src)

let rec mkdirs (pre : string) (path : string) : unit =
  if Filename.concat path "" <> pre then mkdirs pre (Filename.dirname path);
  if not (Sys.file_exists path) then Sys.mkdir path 0o777

let copy_files (out_dir : string) (srcfiles : string list)
    (dstfiles : string list) : unit =
  List.iter
    (fun (src, dst) ->
      mkdirs out_dir (Filename.dirname dst);
      copy_file src dst)
    (List.combine srcfiles dstfiles)

let replace_prefix (old_prefix : string) (new_prefix : string)
    (files : string list) : string list =
  let n = String.length old_prefix in
  List.map
    (fun f -> Filename.concat new_prefix (String.sub f n (String.length f - n)))
    files

let really_alias_files
    ((lib, srcdir, outdir, srcfiles) : string * string * string * string list) :
    unit =
  copy_files outdir srcfiles (replace_prefix srcdir outdir srcfiles)

let really_alias_map
    ((lib, _srcdir, out, srcfiles) : string * string * string * string list) :
    unit =
  let lib = String.capitalize_ascii lib in
  let modules =
    List.sort_uniq String.compare
      (List.rev_map
         (fun f ->
           String.capitalize_ascii
             (Filename.remove_extension (Filename.basename f)))
         srcfiles)
  in
  let content =
    Printf.sprintf "%s\n"
      (String.concat "\n"
         (List.map
            (fun m -> Printf.sprintf "module %s = %s__%s" m lib m)
            modules))
  in
  write_file out content

let command = ref ""
let lib = ref ""
let out = ref ""
let srcdir = ref ""
let srcfiles = ref []

let ensure_trailing_slash (s : string) : string =
  if not (String.ends_with ~suffix:"/" s) then Filename.concat s "" else s

let alias_files () : unit =
  let lib = !lib in
  let srcdir = ensure_trailing_slash !srcdir in
  let out = ensure_trailing_slash !out in
  let srcfiles =
    List.filter
      (fun f -> Sys.file_exists f && String.starts_with ~prefix:srcdir f)
      !srcfiles
  in
  really_alias_files (lib, srcdir, out, srcfiles)

let alias_map () : unit =
  let lib = !lib in
  let srcdir = ensure_trailing_slash !srcdir in
  let out = !out in
  let srcfiles =
    List.filter
      (fun f -> Sys.file_exists f && String.starts_with ~prefix:srcdir f)
      !srcfiles
  in
  really_alias_map (lib, srcdir, out, srcfiles)

let go () : unit =
  match !command with
  | "alias-files" -> alias_files ()
  | "alias-map" -> alias_map ()
  | cmd -> Printf.printf "Unrecognized command: '%s'" cmd

let usage : string =
  {|proto_macro_tool.opt -c <command> [<args>]

   Commands:
   alias-files  Make aliased module files (e.g. copy 'a.ml',.. to 'mylib__A.ml',..).
   alias-map    Generate an alias map ('.mli' file) (e.g. write 'module A = Mylib__A',..).

   Args:|}

let args : (string * Arg.spec * string) list =
  [
    ("-c", Arg.Set_string command, "Command");
    ("-l", Arg.Set_string lib, "Library name");
    ("-s", Arg.Set_string srcdir, "Source root directory");
    ("-o", Arg.Set_string out, "Output file or directory name");
  ]

let files (filename : string) : unit = srcfiles := filename :: !srcfiles

let (_ : unit) =
  Arg.parse args files usage;
  go ()

File: python/proto-macro-tool/proto_macro_tool.py
#!/usr/bin/env python

# format:
#   - autopep8 --in-place --aggressive proto_macro_tool.py

# usage examples:
#   - python3 proto_macro_tool.py alias_files mylib -s ~/tmp/proto-macro-test -f ~/tmp/proto-macro-test/a.mli ~/tmp/proto-macro-test/a.ml ~/tmp/proto-macro-test/b.ml ~/tmp/proto-macro-test/quux/c.ml -o ~/tmp/proto-macro-tool-test-out/
#   - python3 proto_macro_tool.py alias_map mylib -s ~/tmp/proto-macro-test -f ~/tmp/proto-macro-test/a.mli ~/tmp/proto-macro-test/a.ml ~/tmp/proto-macro-test/b.ml ~/tmp/proto-macro-test/quux/c.ml -o ~/tmp/proto-macro-tool-test-out/mylib.mli

import argparse
import os
import shutil
import sys


class ProtoMacroTool(object):
    def __init__(self):
        parser = argparse.ArgumentParser(
            usage='''proto_macro_tool.py <command> [<args>]
Commands:
alias_files    Alias modues (e.g. copy 'a.ml',.. to 'mylib__A.ml',..).
alias_map      Generate an alias map ('.mli' file) (e.g. write 'module A = Mylib__A',..).
''')
        parser.add_argument('command', help='proto_macro_tool.py commands')
        parser.add_argument(
            '-v',
            '--version',
            help='show version and   exit',
            action='version',
            version='1.0')
        args = parser.parse_args(sys.argv[1:2])
        getattr(self, args.command)()

    def alias_files(self):
        parser = argparse.ArgumentParser(description='generate alias modules')
        parser.add_argument('lib', help='lib name e.g. mylib')
        parser.add_argument(
            '-o',
            '--out',
            required=True,
            help='output directory')
        parser.add_argument(
            '-s',
            '--srcdir',
            required=True,
            help='source root directory')
        parser.add_argument(
            '-f',
            '--files',
            required=True,
            nargs='*',
            help='source modules')
        args = parser.parse_args(sys.argv[2:])

        lib = args.lib
        srcdir = os.path.join(args.srcdir, '')
        out = os.path.join(args.out, '')
        files = [f for f in args.files if os.path.isfile(f)]

        cmd_alias_files(lib, srcdir, out, files)

    def alias_map(self):
        parser = argparse.ArgumentParser(
            description='generate modue alias map')
        parser.add_argument('lib', help='lib name e.g. mylib')
        parser.add_argument(
            '-s',
            '--srcdir',
            required=True,
            help='source root directory')
        parser.add_argument(
            '-o',
            '--out',
            required=True,
            help='output map filename (.mli)')
        parser.add_argument(
            '-f',
            '--files',
            required=True,
            nargs='*',
            help='source modules')
        args = parser.parse_args(sys.argv[2:])

        lib = args.lib
        srcdir = os.path.join(args.srcdir, '')
        out = open(args.out, 'w')
        files = [f for f in args.files if os.path.isfile(f)]

        cmd_alias_map(lib, srcdir, out, files)

# --


def cmd_alias_files(lib, srcdir, out, files):
    for src, dst in zip(files, [os.path.join(os.path.dirname(f.replace(
            srcdir, out)), lib + "__" + os.path.basename(f).capitalize()) for f in files]):
        d = os.path.dirname(dst)
        if not os.path.isdir(d):
            os.makedirs(d, exist_ok=True)
        shutil.copyfile(src, dst)
        print("%s -> %s" % (src, dst))


def cmd_alias_map(lib, _srcdir, out, files):
    lib = lib.capitalize()
    modules = {os.path.splitext(os.path.basename(f).capitalize())[
        0]: None for f in files}
    content = ''.join(["module {} = {}__{}\n".format(m, lib, m)
                      for m in modules.keys()])
    out.write(content)
    print("%s" % content)

# --


if __name__ == '__main__':
    ProtoMacroTool()
